Stop propagation on native events

caillou opened the pull request:
### Summary ###

This PR ensures that `event.stopPropagation()` is respected for event handlers that were registered using the native `element.addEventlistener()` method when the event was triggered by jQuery's `.trigger('click')` method.

This PR fixes #3693.

The ticket was closed at first with the following argument:

> jQuery has never been able to guarantee full interoperability of native-vs-jQuery event handlers.

Nonetheless, the discussion lead me to believe, that this incompatibility between native and jQuery event handling could be fixed and that such a fix is worthwhile.

### Checklist ###

* [x] All authors have signed the CLA at https://cla.js.foundation/jquery/jquery
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* [x] If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com

dmethvin commented on src/event/trigger.js:
```diff
@@ -10,13 +10,16 @@ define( [
 
-var rfocusMorph = /^(?:focusinfocus|focusoutblur)$/;
+var rfocusMorph = /^(?:focusinfocus|focusoutblur)$/,
+	stopPropagationCallback = function( e ) {
```
I wonder if we should have this inside `trigger()` since it is only used there. Opinions welcome.

caillou replied:
The difference is the following:

* If we define the function within the `trigger` method, it is defined every time trigger is called.
* If we define the function within the AMD callback function as I did here, the function will be defined only once, but even if the `trigger` method is never used. Additionally, there is an additional scope lookup.

All in all, these considerations are micro optimisations, that are not really worth considering. The questions is more one of culture within a project. We could also define it at the end of the file, just before the `return jQuery` statement.

I will look at the other files and see how it is usually handled within the `jQuery` project.

dmethvin commented on src/event/trigger.js:
```diff
@@ -136,7 +139,17 @@ jQuery.extend( jQuery.event, {
 					jQuery.event.triggered = type;
+
+					if ( event.isPropagationStopped() ) {
+						lastElement.addEventListener( type, stopPropagationCallback );
```
I think it would be possible to get here with `lastElement === undefined` if the incoming `event` already had propagation stopped? Maybe initialize with `lastElement = elem`?

dmethvin replied:
Hmmm, this does bring up the problem of `.stopImmediatePropagation()` which isn't fixed. Since that method is sensitive to the order that handlers are attached and ours is always going to be last, there isn't much we can do for that 🤷‍♂️ .

caillou replied:
@dmethvin One option with the `.stopImmediatePropagation()` would be to register an additional event handler on the second to last element (the great ancestor) if there is one present, that stops the propagation.

But this seems rather fiddly.

caillou replied:
@dmethvin I will try to add a test-case for the `lastElement === undefined` case. But I don't yes see how this could happen. How can the propagation of the incoming `even` already be stopped?

dmethvin commented on test/unit/event.js:
```diff
@@ -440,6 +440,26 @@ QUnit.test( "on bubbling, isDefaultPrevented, stopImmediatePropagation", functio
 	$anchor2[ 0 ].removeEventListener( "click", neverCallMe );
 } );
 
+QUnit.test( "triggerd events stopPropagation() for natively-bound events", function( assert ) {
```
triggered

gibson042 commented on src/event/trigger.js:
```diff
@@ -93,7 +96,7 @@ jQuery.extend( jQuery.event, {
 		i = 0;
 		while ( ( cur = eventPath[ i++ ] ) && !event.isPropagationStopped() ) {
-
+			lastElement = cur;
```
You can reduce the size increase of jquery.min.gz by using `eventPath[ i - 2 ]` below rather than introducing a new variable.

caillou replied:
@gibson042 in my opinion, it would reduce the size at a high cost of code readability. What is the code guideline in the `jQuery` project with regard to readability v.s. size?

caillou commented:
@dmethvin @gibson042: I have added a test with a perviously stopped event and fixed it.

Could you review this again? Sorry for the trouble, I should have thought about this.

This yields the question: Do we really want to go forward with this PR? While it helps with interoperability between native and jQuery events, I wonder if it is worth the trouble. WDYT?

dmethvin commented:
@caillou I think you have a pretty small patch here and it could be useful. What are your concerns?

caillou commented:
@dmethvin Good. No real concerns. What is the workflow? Do I need to do anything, or will it get merged at some point?

dmethvin commented:
@caillou Yes, we'll merge it shortly. Thanks!
